### PR TITLE
Raise error on `nan` in `_constrained_dominates`

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
Currently, NSGA-II behaves very counterintuitively when `nan` is given as constraint values. `nan` values are treated as infeasible, but it is currently treated as zero violation.

We resolve this issue by simply raising `ValueError` when a `nan` is found.

## Description of the changes
Raise `ValueError` in `_constrained_dominates` function when a `nan` constraint is encountered.

### CODE REVIEW

**Comment by contramundum53** [comment]
I made a simple tracking issue -> #3746

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Comment by HideakiImamura** [comment]
@keisuke-umezawa Could you review this PR?

File: optuna/samplers/nsgaii/_sampler.py [hunk 1]
```diff
@@ -485,6 +485,9 @@ def _constrained_dominates(
     assert isinstance(constraints0, (list, tuple))
     assert isinstance(constraints1, (list, tuple))
 
+    if np.any(np.isnan(constraints0)) or np.any(np.isnan(constraints1)):
+        raise ValueError("NaN is not acceptable as constraint value.")
```
Comment (keisuke-umezawa) [inline optuna/samplers/nsgaii/_sampler.py:489]: [question] For example, only constraints1 has nan, can we assume that constraints0 is feasible? like follows

```
    if constraints0 is not None and constraints1 is None:
        # Trial x has constraint values, but y doesn't.
        return True
```

Comment (contramundum53) [reply optuna/samplers/nsgaii/_sampler.py:489]: I think having `nan` in constraints means there is a bug in the user code, and we should notify the user as early as possible. That's why I didn't consider `nan` constraint as an infeasible trial but reported error.

**Comment by HideakiImamura** [comment]
@keisuke-umezawa is busy now, so could you review this PR? @nzw0301

**Review by nzw0301 — COMMENTED** [review]
I think we don't need `_nan_equal` anymore because this PR drops `nan` from the parameterization for the constraint. Could you remove it for readability?
https://github.com/optuna/optuna/blob/6c643ea342ff9fb2dc85fc4c9e5f380871968eaa/tests/samplers_tests/test_nsgaii.py#L156

File: tests/samplers_tests/test_nsgaii.py [hunk 2]
```diff
@@ -157,6 +156,15 @@ def constraints_func(trial: FrozenTrial) -> Sequence[float]:
             assert _nan_equal(x, y)
```
Comment (contramundum53) [inline tests/samplers_tests/test_nsgaii.py:156]:
```suggestion
            assert x == y
```

**Comment by contramundum53** [comment]
Fixed. PTAL. @nzw0301

**Review by nzw0301 — APPROVED** [review]
Look great. Thanks!

**Pull request merged**